File: VB/SnxToRepx/Converter/SnapListCollection.vb
Imports DevExpress.Snap.Core.API
Imports DevExpress.XtraRichEdit.API.Native
Imports System.Collections.Generic

Namespace SnxToRepx.Converter

    ''' <summary>
    ''' This is a collection of SnapList fields with additional logic
    ''' </summary>
    Friend Class SnapListCollection
        Inherits List(Of Field)

        Private source As SnapDocument 'The source document

        Public Sub New(ByVal source As SnapDocument)
            Me.source = source
        End Sub

        ''' <summary>
        ''' This method collects top-level SnapLists
        ''' </summary>
        Public Sub PrepareCollection()
            'Clear the collection
            Clear()
            'Iterate through fields
            For Each field As Field In source.Fields
                'Get a top-level field for the current field
                Dim topField As Field = GetTopLevelField(field)
                If Not ContainsField(topField) Then
                    'If the field is not yet added and it is a SnapList, add it to the collection
                    Dim entity As SnapEntity = source.ParseField(topField)
                    If entity IsNot Nothing AndAlso TypeOf entity Is SnapList Then Add(topField)
                End If
            Next

            'Sort lists based on their ranges (SnapLists can be in the reverse order in the document model)
            Sort(New FieldComparer())
        End Sub

        ''' <summary>
        ''' Gets a SnapList instance based on its index in the collection
        ''' </summary>
        ''' <param name="index"></param>
        ''' <returns></returns>
        Public Function GetSnapList(ByVal index As Integer) As SnapList
            'Get and parse the source field
            Dim field As Field = Me(index)
            Dim entity As SnapEntity = source.ParseField(field)
            'Return SnapList
            Return CType(entity, SnapList)
        End Function

        ''' <summary>
        ''' Gets a SnapList instance containing the specified position
        ''' </summary>
        ''' <param name="position">A position to check</param>
        ''' <returns>The SnapList</returns>
        Public Function GetSnapListByPosition(ByVal position As Integer) As SnapList
            'Iterate through elements in the collection
            For i As Integer = 0 To Count - 1
                'Get a field and compare its range with the specified position
                Dim field As Field = Me(i)
                'If they match, return SnapList
                If field.Range.Start.ToInt() <= position AndAlso field.Range.End.ToInt() >= position Then Return CType(source.ParseField(field), SnapList)
            Next

            'If no SnapList is found, return Null
            Return Nothing
        End Function

        ''' <summary>
        ''' Gets the top-level field for the current field
        ''' </summary>
        ''' <param name="field">The currently processed field</param>
        ''' <returns>The top-level field</returns>
        Private Function GetTopLevelField(ByVal field As Field) As Field
            'Access the parent field
            Dim parentField As Field = field
            While parentField.Parent IsNot Nothing
                parentField = parentField.Parent
            End While

            'Return the resulting field
            Return parentField
        End Function

        ''' <summary>
        ''' Checks whether the collection contains a specific field
        ''' </summary>
        ''' <param name="field">A field to locate</param>
        ''' <returns>True if the collection contains the field. Otherwise, false.</returns>
        Public Function ContainsField(ByVal field As Field) As Boolean
            'Iterate through fields in the collection
            For i As Integer = 0 To Count - 1
                'Get a field and compare its range with the specified position
                Dim addedField As Field = Me(i)
                If addedField.Range.Start.ToInt() = field.Range.Start.ToInt() AndAlso addedField.Range.Start.ToInt() = field.Range.Start.ToInt() Then Return True
            Next

            Return False
        End Function
    End Class

    ''' <summary>
    ''' A comparer used to sort fields by their positions
    ''' </summary>
    Friend Class FieldComparer
        Implements IComparer(Of Field)

        Public Function Compare(ByVal x As Field, ByVal y As Field) As Integer Implements IComparer(Of Field).Compare
            'Use the default Integer comparer to compare positions
            Return Comparer(Of Integer).Default.Compare(x.Range.Start.ToInt(), y.Range.Start.ToInt())
        End Function
    End Class
End Namespace

File: CS/SnxToRepx/Converter/SnapListCollection.cs
using DevExpress.Snap.Core.API;
using DevExpress.XtraRichEdit.API.Native;
using System.Collections.Generic;

namespace SnxToRepx.Converter
{
    /// <summary>
    /// This is a collection of SnapList fields with additional logic
    /// </summary>
    class SnapListCollection : List<Field>
    {
        private SnapDocument source; //The source document
        public SnapListCollection(SnapDocument source)
        {
            this.source = source;
        }

        /// <summary>
        /// This method collects top-level SnapLists
        /// </summary>
        public void PrepareCollection()
        {
            //Clear the collection
            this.Clear();

            //Iterate through fields
            foreach (Field field in source.Fields)
            {
                //Get a top-level field for the current field
                Field topField = GetTopLevelField(field);
                if (!this.ContainsField(topField))
                {
                    //If the field is not yet added and it is a SnapList, add it to the collection
                    SnapEntity entity = source.ParseField(topField);
                    if (entity != null && entity is SnapList)
                        this.Add(topField);
                }
            }

            //Sort lists based on their ranges (SnapLists can be in the reverse order in the document model)
            this.Sort(new FieldComparer());
        }

        /// <summary>
        /// Gets a SnapList instance based on its index in the collection
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public SnapList GetSnapList(int index)
        {
            //Get and parse the source field
            Field field = this[index];
            SnapEntity entity = source.ParseField(field);
            //Return SnapList
            return (SnapList)entity;
        }

        /// <summary>
        /// Gets a SnapList instance containing the specified position
        /// </summary>
        /// <param name="position">A position to check</param>
        /// <returns>The SnapList</returns>
        public SnapList GetSnapListByPosition(int position)
        {
            //Iterate through elements in the collection
            for (int i = 0; i < this.Count; i++)
            {
                //Get a field and compare its range with the specified position
                Field field = this[i];
                if (field.Range.Start.ToInt() <= position &&
                    field.Range.End.ToInt() >= position)
                    //If they match, return SnapList
                    return (SnapList)source.ParseField(field);
            }
            //If no SnapList is found, return Null
            return null;
        }

        /// <summary>
        /// Gets the top-level field for the current field
        /// </summary>
        /// <param name="field">The currently processed field</param>
        /// <returns>The top-level field</returns>
        private Field GetTopLevelField(Field field)
        {
            //Access the parent field
            Field parentField = field;
            while (parentField.Parent != null)
                parentField = parentField.Parent;

            //Return the resulting field
            return parentField;
        }   
        
        /// <summary>
        /// Checks whether the collection contains a specific field
        /// </summary>
        /// <param name="field">A field to locate</param>
        /// <returns>True if the collection contains the field. Otherwise, false.</returns>
        public bool ContainsField(Field field)
        {
            //Iterate through fields in the collection
            for (int i = 0; i < this.Count; i++)
            {
                //Get a field and compare its range with the specified position
                Field addedField = this[i];
                if (addedField.Range.Start.ToInt() == field.Range.Start.ToInt() &&
                    addedField.Range.Start.ToInt() == field.Range.Start.ToInt())
                    return true;
            }
            return false;
        }
    }

    /// <summary>
    /// A comparer used to sort fields by their positions
    /// </summary>
    class FieldComparer : IComparer<Field>
    {
        public int Compare(Field x, Field y)
        {
            //Use the default Integer comparer to compare positions
            return Comparer<int>.Default.Compare(x.Range.Start.ToInt(), y.Range.Start.ToInt());
        }
    }
}
